### PR TITLE
This PR is adding discount coupon feature to vendor 

### LINKED ISSUE #89: Vendor should be able to manage discount coupons
# Description

The backend needs to handle the creation, management, and application of coupon codes, including storing the coupon code information in the database and applying discounts to eligible products during checkout. The system should also allow vendors to retrieve and view their associated coupon codes.

## Acceptance Criteria:

- [ ]  GIVEN a post request to create a new coupon code
- [ ] WHEN the requesting user/token is a vendor and the request body is validated
- [ ] THEN the new coupon code should be saved to the database with the vendor's account ID, discount amount/percentage, expiration date, maximum usage limit, and associated products/product groups.
- [ ]  GIVEN a post request to apply a coupon code during checkout
- [ ] GIVEN a post request to get coupon codes

## Dev Notes:

- The coupon code model should include fields for discount amount/percentage, expiration date, maximum usage limit, associated products/product groups, and the vendor's account ID.
- Implement appropriate error handling and validation mechanisms to ensure data integrity and prevent misuse of coupon codes.
- Consider implementing additional features such as coupon code management (editing, disabling, deleting)
- Ensure proper access control and authorization mechanisms are in place to prevent unauthorized access or modification of coupon codes.

## Definition of Done:

- Vendor is able to manage discount coupons
- Code review has been completed, and any necessary changes have been addressed.
- Documentation for the coupon code feature has been updated.
- Test are passing.

### PR DESCRIPTION
**Pull request by Ndevu12**
# Description

This PR introduces the discount coupon feature, allowing vendors to create, manage, and apply discount coupons to customer orders. Key changes include database schema updates, new API endpoints, and comprehensive testing.

### Database Design

- Created a coupons table with columns for ID, code, discount type/value, expiration date, usage counts, vendor ID, product IDs, and timestamps.
- Updated vendors, customers, orders, and customerCoupons tables for relationships with coupons.

### API Endpoints

- Create Coupon: POST /api/coupons
- Apply Coupon: POST /api/checkout/apply-coupon
- Get Vendor Coupons: GET /api/vendors/:vendorId/coupons

### Testing

- Added unit tests for coupon functionality.
- Added integration tests for API endpoints.

### Documentation

- Updated API documentation with new coupon endpoints.

### CODE REVIEW

**Comment by faid-terence** [comment]
@Ndevu12 
This is not a proper PR Title , please rename it

**Comment by Ndevu12** [comment]
@faid-terence renamed

**Review by Calebgisa72 — COMMENTED** [review]
In the coupon entity check on adding a field to keep track of the number of times the coupon has been used, it can start by zero and be incremented by one as a buyer use that coupon

**Comment by faid-terence** [comment]
Screenshot from 2024-05-24 15-17-39
@Ndevu12 Why did you delete these tests ??

File: src/__test__/cart.test.test.ts
Comment (faid-terence) [inline src/__test__/cart.test.test.ts:1]: @Ndevu12 
Why did you deleted these tests ?

File: src/__test__/getProduct.test.ts
Comment (faid-terence) [inline src/__test__/getProduct.test.ts:1]: @Ndevu12 
Why did you deleted these tests ?

File: src/__test__/userStatus.test.ts
Comment (faid-terence) [inline src/__test__/userStatus.test.ts:1]: @Ndevu12 
Why did you deleted these tests ?

File: src/__test__/vendorProduct.test.ts
Comment (faid-terence) [inline src/__test__/vendorProduct.test.ts:1]: @Ndevu12 
Why did you deleted these tests ?

File: src/__test__/wishList.test.ts
Comment (faid-terence) [inline src/__test__/wishList.test.ts:1]: @Ndevu12 
Why did you deleted these tests ?

**Review by faid-terence — CHANGES_REQUESTED** [review]
@Ndevu12 Please get all tests back , you deleted them !!!

**Comment by Ndevu12** [comment]
> @Ndevu12 Please get all tests back , you deleted them !!!

@faid-terence Sorry for this mistake. I can't say that I didn't do it by intention. The reason for the deletion of these tests is that I was running out of time and I had been fixing the errors they had around 30 minutes. so, because I had an exam at 3:00PM. I had to do quick things so, that I can attend the exam and all others who depend on this PR can progress on their projects. If needed, I can work the tests. Am ready. Also I have managed to write API endpoints documentation

**Comment by faid-terence** [comment]
@Ndevu12 Understood but never do that again , it's selfish removing other people's work to keep yours 
Please work together as a team !

**Comment by Ndevu12** [comment]
> @Ndevu12 Understood but never do that again , it's selfish removing other people's work to keep yours 
> Please work together as a team !
> 
> 

- @faid-terence noted Sir, won't happen again, but I was trying to not keeping others not working on their works. But before any other things I will ask for a permission even if it will be urgent. Thanks for your understanding. Best regards.

**Comment by Ndevu12** [comment]
> @Ndevu12 Please get all tests back , you deleted them !!!

@faid-terence Now, All tests are added back!

**Review by aimedivin — APPROVED** [review]

**Review by faid-terence — APPROVED** [review]
PR Approved @Ndevu12 !

**Comment by Ndevu12** [comment]
> PR Approved @Ndevu12 !

@faid-terence thanks for the assistance and awareness!

**Review by MC-Knight — APPROVED** [review]

**Review by UwicyezaG — APPROVED** [review]

**Review by aimedivin — APPROVED** [review]

**Review by Calebgisa72 — APPROVED** [review]

**Pull request merged**